File: mobile/android/app/src/main/java/net/matrixcapital/vitaforge/core/ApiClient.kt
package net.matrixcapital.vitaforge.core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import net.matrixcapital.vitaforge.model.TokenPair
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

/**
 * OkHttp + kotlinx.serialization networking. JWT auth with **single-flight token
 * refresh** on 401 (a Mutex coalesces concurrent refreshes), and Accept-Language
 * on every request. Inline reified helpers keep call sites terse.
 */
class ApiClient(
    private val tokenStore: TokenStore,
    private val acceptLanguage: () -> String,
) {
    @PublishedApi
    internal val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        explicitNulls = false
    }

    private val http = OkHttpClient()
    private val refreshMutex = Mutex()
    private val mediaType = "application/json".toMediaType()

    // ---- typed helpers ----

    suspend inline fun <reified T> get(path: String, query: Map<String, String>? = null): T =
        json.decodeFromString(perform(path, "GET", null, authed = true, query = query))

    suspend inline fun <reified B, reified T> post(path: String, body: B): T =
        json.decodeFromString(perform(path, "POST", json.encodeToString(body), authed = true, query = null))

    suspend inline fun <reified B, reified T> put(path: String, body: B): T =
        json.decodeFromString(perform(path, "PUT", json.encodeToString(body), authed = true, query = null))

    suspend fun delete(path: String) {
        perform(path, "DELETE", null, authed = true, query = null)
    }

    /** Unauthenticated call (register/login/guest preview). */
    suspend inline fun <reified B, reified T> anonymous(path: String, body: B): T =
        json.decodeFromString(perform(path, "POST", json.encodeToString(body), authed = false, query = null))

    // ---- core ----

    private data class Raw(val code: Int, val body: String)

    @PublishedApi
    internal suspend fun perform(
        path: String,
        method: String,
        body: String?,
        authed: Boolean,
        query: Map<String, String>?,
    ): String = withContext(Dispatchers.IO) {
        val access = if (authed) tokenStore.access else null
        var resp = rawCall(path, method, body, access, query)

        if (resp.code == 401 && authed) {
            try {
                val fresh = refreshTokens(staleAccess = access)
                resp = rawCall(path, method, body, fresh, query)
            } catch (e: ApiException) {
                tokenStore.clear()
                throw ApiException.Unauthorized()
            }
        }

        if (resp.code !in 200..299) {
            throw ApiException.Server(resp.code, detailOf(resp.body, resp.code))
        }
        resp.body
    }

    /** Single-flight: concurrent 401s wait on the mutex; whoever wins refreshes,
     *  the rest see the already-rotated access token and skip a second refresh. */
    private suspend fun refreshTokens(staleAccess: String?): String = refreshMutex.withLock {
        val current = tokenStore.access
        if (current != null && current != staleAccess) return current

        val refresh = tokenStore.refresh ?: throw ApiException.Unauthorized()
        val payload = buildJsonObject { put("refresh_token", refresh) }.toString()
        val resp = rawCall("/auth/refresh", "POST", payload, accessToken = null, query = null)
        if (resp.code !in 200..299) throw ApiException.Unauthorized()
        val pair = json.decodeFromString<TokenPair>(resp.body)
        tokenStore.save(pair)
        pair.accessToken
    }

    private fun rawCall(
        path: String,
        method: String,
        body: String?,
        accessToken: String?,
        query: Map<String, String>?,
    ): Raw {
        val url = (ApiConfig.BASE_URL + path).toHttpUrl().newBuilder().apply {
            query?.forEach { (k, v) -> addQueryParameter(k, v) }
        }.build()

        val requestBody = when {
            body != null -> body.toRequestBody(mediaType)
            method == "POST" || method == "PUT" -> "".toRequestBody(mediaType)
            else -> null
        }

        val builder = Request.Builder()
            .url(url)
            .header("Accept-Language", acceptLanguage())
            .header("Content-Type", "application/json")
            .method(method, requestBody)
        if (accessToken != null) builder.header("Authorization", "Bearer $accessToken")

        return try {
            http.newCall(builder.build()).execute().use { r ->
                Raw(r.code, r.body?.string() ?: "")
            }
        } catch (e: IOException) {
            throw ApiException.Network(e)
        }
    }

    /** Pull a human message from a FastAPI `{ "detail": … }` body. */
    private fun detailOf(body: String, status: Int): String {
        return try {
            val obj = json.parseToJsonElement(body).jsonObject
            val detail = obj["detail"]
            when {
                detail == null -> "Request failed ($status)."
                detail.toString().startsWith("[") ->
                    detail.jsonArray.firstOrNull()?.jsonObject?.get("msg")?.jsonPrimitive?.content
                        ?: "Request failed ($status)."
                else -> detail.jsonPrimitive.content
            }
        } catch (e: Exception) {
            "Request failed ($status)."
        }
    }
}

File: mobile/ios/Sources/Core/APIClient.swift
import Foundation

/// Async networking layer. An `actor` so token refresh is single-flight: many
/// concurrent 401s coalesce onto one `/auth/refresh` (MOBILE_API.md). JSON is
/// snake_case on the wire; models stay camelCase via key strategies.
actor APIClient {
    static let shared = APIClient()

    private let session: URLSession
    private let tokens = KeychainTokenStore.shared
    private var refreshTask: Task<TokenPair, Error>?

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()
    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        return e
    }()

    init(session: URLSession = .shared) { self.session = session }

    // MARK: - Typed helpers

    func get<R: Decodable>(_ path: String, query: [String: String]? = nil) async throws -> R {
        try decoder.decode(R.self, from: await perform(path, method: "GET", body: nil, authed: true, query: query))
    }

    func post<B: Encodable, R: Decodable>(_ path: String, body: B) async throws -> R {
        try decoder.decode(R.self, from: await perform(path, method: "POST", body: try encoder.encode(body), authed: true, query: nil))
    }

    func put<B: Encodable, R: Decodable>(_ path: String, body: B) async throws -> R {
        try decoder.decode(R.self, from: await perform(path, method: "PUT", body: try encoder.encode(body), authed: true, query: nil))
    }

    /// POST/PUT/DELETE that returns no body (204) or whose body we ignore.
    @discardableResult
    func send<B: Encodable>(_ path: String, method: String, body: B) async throws -> Data {
        try await perform(path, method: method, body: try encoder.encode(body), authed: true, query: nil)
    }

    @discardableResult
    func send(_ path: String, method: String) async throws -> Data {
        try await perform(path, method: method, body: nil, authed: true, query: nil)
    }

    /// Unauthenticated call (guest preview, register, login).
    func anonymous<B: Encodable, R: Decodable>(_ path: String, method: String = "POST", body: B) async throws -> R {
        try decoder.decode(R.self, from: await perform(path, method: method, body: try encoder.encode(body), authed: false, query: nil))
    }

    // MARK: - Core

    private func perform(_ path: String, method: String, body: Data?, authed: Bool, query: [String: String]?) async throws -> Data {
        let request = buildRequest(path, method: method, body: body, query: query, accessToken: authed ? tokens.cached?.accessToken : nil)
        let (data, status) = try await raw(request)

        // One transparent refresh-and-retry on 401 for authed calls.
        if status == 401, authed {
            do {
                let pair = try await refreshTokens()
                let retry = buildRequest(path, method: method, body: body, query: query, accessToken: pair.accessToken)
                let (retryData, retryStatus) = try await raw(retry)
                guard (200..<300).contains(retryStatus) else {
                    throw APIError.server(status: retryStatus, detail: APIError.detail(from: retryData, status: retryStatus))
                }
                return retryData
            } catch {
                tokens.clear()
                throw APIError.unauthorized
            }
        }

        guard (200..<300).contains(status) else {
            throw APIError.server(status: status, detail: APIError.detail(from: data, status: status))
        }
        return data
    }

    private func raw(_ request: URLRequest) async throws -> (Data, Int) {
        do {
            let (data, response) = try await session.data(for: request)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            return (data, status)
        } catch {
            throw APIError.network(error)
        }
    }

    private func buildRequest(_ path: String, method: String, body: Data?, query: [String: String]?, accessToken: String?) -> URLRequest {
        var url = APIConfig.baseURL.appendingPathComponent(path.hasPrefix("/") ? String(path.dropFirst()) : path)
        if let query, var comps = URLComponents(url: url, resolvingAgainstBaseURL: false) {
            comps.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
            url = comps.url ?? url
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(AppLanguage.current.rawValue, forHTTPHeaderField: "Accept-Language")
        if let accessToken { req.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization") }
        req.httpBody = body
        return req
    }

    // MARK: - Refresh (single-flight)

    private func refreshTokens() async throws -> TokenPair {
        if let existing = refreshTask { return try await existing.value }
        let task = Task<TokenPair, Error> { [self] in try await doRefresh() }
        refreshTask = task
        defer { refreshTask = nil }
        return try await task.value
    }

    private func doRefresh() async throws -> TokenPair {
        guard let refresh = tokens.cached?.refreshToken else { throw APIError.unauthorized }
        let body = try encoder.encode(["refresh_token": refresh])
        let req = buildRequest("/auth/refresh", method: "POST", body: body, query: nil, accessToken: nil)
        let (data, status) = try await raw(req)
        guard (200..<300).contains(status) else { throw APIError.unauthorized }
        let pair = try decoder.decode(TokenPair.self, from: data)
        tokens.save(pair)
        return pair
    }
}
